Floating IPs provisioning UI

gildub opened the pull request:
Adds controller actions using task queue
Adds new and edit views

https://bugzilla.redhat.com/show_bug.cgi?id=1394277

tzumainn commented:
@miq-bot add_label euwe/yes

gildub commented:
@martinpovolny, @himdel, @dclarizio, can you please review?

himdel commented:
@gildub Can you please update the PR description with instructions where to find this functionality and a screenshot?

tzumainn commented:
![screenshot from 2016-11-07 10 57 18](https://cloud.githubusercontent.com/assets/1280178/20064512/09e45d76-a4d9-11e6-977a-945e85e440eb.png)
![screenshot from 2016-11-07 10 57 23](https://cloud.githubusercontent.com/assets/1280178/20064513/09e8c2da-a4d9-11e6-9f6b-1442380133cd.png)

tzumainn commented:
Accessed through the left navigation - Networks > Floating IPs

tzumainn commented:
@himdel added the info, let me know if it helps!  I've tested this functionality successfully, so I think we're only really looking for code sanity.

himdel commented on app/controllers/floating_ip_controller.rb:
```diff
@@ -14,4 +14,262 @@ def self.display_methods
+
+    floating_ips_to_delete = []
+    floating_ips.each do |s|
+      floating_ip = FloatingIp.find_by_id(s)
```
This should be `find(s)`

gildub replied:
Done

himdel commented:
> Floating IP Address (optional)

So.. I'm a bit concerned about the address field being optional. In particular, @mansam has done changes to the UI in https://github.com/ManageIQ/manageiq/pull/12446 that depend on every floating ip having an `address`.

So how is this supposed to work? Is that a valid expectation, do the addresses get filled out during the task or something like that, or do we need that field to be required? (Sorry, I know nothing..)

himdel commented:
Another thing, trying to save a new floating IP, I get an error that just says "Floating IP creation failed: Task start failed: ID [10000000088263]".

That's ..very unhelpful :) (What I'm seeing in the log is..., so we do have an actual error message we could show..)

```
[----] E, [2016-11-10T13:33:41.336507 #14977:2b0527c090d4] ERROR -- : excon.error     #<Excon::Error::Unauthorized: Expected([200, 204]) <=> Actual(401 Unauthorized)
excon.error.response
  :body          => "{\"error\": {\"message\": \"The request you have made requires authentication.\", \"code\": 401, \"title\": \"Unauthorized\"}}"
  :cookies       => [
  ]
```

gildub commented:
@himdel,

A floating IP address from the pool would be automatically assigned unless provided.

I fixed the task error.

gildub commented on app/controllers/floating_ip_controller.rb:
```diff
@@ -14,4 +14,262 @@ def self.display_methods
+  def floating_ip_form_fields
+    assert_privileges("floating_ip_edit")
+    # TODO: find_by_id_filtered fails
+    # floating_ip = find_by_id_filtered(FloatingIp, params[:id])
```
Fixed

himdel replied:
You say fixed, but I still see the commented out code here..

himdel replied:
Btw seems like this is working fine in SecurityGroups so maybe it would now here too..?

tzumainn commented on app/controllers/floating_ip_controller.rb:
```diff
@@ -14,4 +14,260 @@ def self.display_methods
+    options[:ems_id] = params[:ems_id] if params[:ems_id] && params[:ems_id] != 'new'
+    options[:floating_ip_address] = params[:floating_ip_address] if params[:floating_ip_address]
+    options[:cloud_network_id] = params[:cloud_network_id] if params[:cloud_network_id]
+    options[:cloud_tenant_id] = params[:cloud_tenant_id] if params[:cloud_tenant_id]
```
This needs to be something like:

options[:cloud_tenant] = find_by_id_filtered(CloudTenant, params[:cloud_tenant_id]) if params[:cloud_tenant_id]

Because the model expects options to include :cloud_tenant, and not :cloud_tenant_id (I'm currently fixing this in other places)

martinpovolny commented on app/controllers/floating_ip_controller.rb:
```diff
@@ -14,4 +14,260 @@ def self.display_methods
 
   menu_section :net
+
+  def button
```
Can you cleanup this method a bit?

Please, make it a simple case statement and make sure that all the code is needed. E.g. you are setting the `@refresh_div` variable and then testing it. Is there some code between the statements that might change that value?

gildub replied:
@martinpovolny,

All the network UI and others (cloud_tenant, cloud_volumes) use the exact same approach.
So I'm keen to make a refactor on all of them using a consistent approach.

That said, I don't see the point of a case for replacing a if/elsif. 
No idea about the @refresh_div, I haven't digged into it, what's for? What's the issue here?
Thanks

himdel commented:
So.. another problem, similar to https://github.com/ManageIQ/manageiq/pull/12101#issuecomment-263303329

1. Go to `/cloud_tenant/show/10000000000011?display=floating_ips` (assuming such a tenant exists) - Compute > Clouds > Tenants - click a tenant detail, click on floating ips in the summary detail.  
OR  
Go to `/network_port/show/10000000000389?display=floating_ips` (same) - Networks > Network Ports - click on a port detail, click on floating ips in the summary detail.

1. pick any ip (checkbox)
1. do any of the new toolbar actions
1. `No route matches {:action=>"layouts/flash_msg", :controller=>"network_port", :id=>nil} [network_port/button]` (or similar)

gildub commented:
@himdel, 

I can see that issue.
I believe this is respectively a cloud tenant and a network_port controller issues because other controllers accessing the same page are responsible for it, isn't ?
That said I'm looking into it.

gildub commented:
@himdel,

Ok, so from the UI endpoint, why when selecting the floating IP (whether from the cloud tenant or network port entry) we don't end up in the floating IP controller?

A "normal" work flow would be:
After the user selects the floating IP, having saved where the user is coming from to be able to go back if needed later, then the floating ip controller manages any changes, ultimately the user chooses to go back or go somewhere else.

But obviously we are on a different road, where a controller (such as cloud tenant or network port, from the example above) is managing the views it's not designed for. Unless there is some logic in those button I'm missing. If that the case could you please direct me to the related documentation or at least some good practice example.

Thanks

gildub commented:
@dclarizio, any pointers please?

himdel commented:
@gildub exactly, the solution is essentially to handle the button action in both places, redirecting to the right controller from the other one. Or sharing the code.

For example, for the edit action, just redirecting to the right URL is usually enough (except you may need something like `javascript_redirect previous_breadcrumb_url` when the form is saved and done (or cancelled)) so that you get back to the right place again.

As @martinpovolny said, we're currently in the process of refactoring this so .. it has to work .. but doesn't have to be perfect right now :).

gildub commented:
@himdel, 

It seems like this is an on-going discussion and the shared controller approach might not be the one to take.

@martinpovolny also suggested to minimize the impact by not having some button appear on the cross controller. Do you have details about how to do that. Sorry but I'm only a button consumer :)

gildub commented:
@himdel, 

It seems like this is an on-going discussion and the shared controller approach might not be the one to take.

@martinpovolny also suggested to minimize the impact by not having some button appear on the cross controller. Do you have details about how to do that?

himdel commented:
> also suggested to minimize the impact by not having some button appear on the cross controller. Do you have details about how to do that?

Can't make them not appear in one controller but yes in another. But what you can do is not add the button to the list toolbar but only to the detail toolbar. So they would not be present in either of the lists, but only in the detail screen.

And that is easy because those are separate toolbars (one is in plural, the other not) .. so you just not add it to the other one :).

> shared controller approach might not be the one to take.

I was not proposing a shared controller, I was proposing exactly the opposite - redirecting those actions from the bad controller to the right one. (Similar to https://github.com/ManageIQ/manageiq/pull/12551/files#diff-68512128862e16616d93240055cb3a7dR77)

gildub commented:
@himdel, 

Oh sorry, I misunderstood you! Effectively #12551 is taking a redirect approach, that's music to me, well once the tricky filtering part is finished.

So I removed the buttons from the show list. I was hoping there was some flag available to do it in order to avoid loosing the feature. But that's okay until we have the cross controller issue addressed.

I've also replaced the button statements with a case as now that I can see a better road path with the buttons.

himdel commented on app/controllers/floating_ip_controller.rb:
```diff
@@ -14,4 +14,265 @@ def self.display_methods
+    render :json => {
+      :fixed_ip_address     => floating_ip.fixed_ip_address,
+      :floating_ip_address  => floating_ip.address,
+      :cloud_network_name   => floating_ip.cloud_network.name,
```
Equally problematic, apparently not all cloud floating ips have a network ;).

And the same for cloud tenant.

I guess `floating_ip.cloud_network.try(:name)`..?

martinpovolny replied:
Again, this seems to be the last bit...

tzumainn commented:
Associated BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1394277

gildub commented:
@miq-bot add_label providers/openstack/cloud

gildub commented:
@himdel, ok now that the button filtering is in place tests are failing with Google and Azure getting 500 error.

How are we supposed to handle other providers? Shouldn't they be dealing with their own stuff?

@tzumain, FYI

himdel commented:
I think the same answer as in https://github.com/ManageIQ/manageiq/pull/13165#issuecomment-267020564 applies here.. just make sure the method returns false for the other providers.

The test failure is again

```
       undefined method `is_available?' for #<FloatingIp:0x0055fd9bf90f08>
       undefined method `supports_update_floating_ip?' for #<FloatingIp:0x0055fd9bf90f08>
```

gildub commented:
@himdel, moving the supportfeaturesmixin on the base class helps

himdel commented on app/controllers/floating_ip_controller.rb:
```diff
@@ -14,4 +14,267 @@ def self.display_methods
+    @in_a_form = true
+    drop_breadcrumb(
+      :name => _("Associate Floating IP \"%{name}\"") % { :name  => @floating_ip.name },
+      :url  => "/floating_ip/edit/#{@floating_ip.id}git")
```
`/edit/#{@floating_ip.id}git`

I'm guessing `git` is not supposed to be there? :)

gildub replied:
Not at all! Thanks

himdel commented on app/assets/javascripts/controllers/floating_ip/floating_ip_form_controller.js:
```diff
@@ -0,0 +1,57 @@
+  } else {
+    miqService.sparkleOn();
+
+    $http.get('/floating_ip/floating_ip_form_fields/' + floatingIpFormId).success(function(data) {
```
Doesn't matter now, I'll be doing a PR to cover the whole codebase anyway, but in the future, please prefer `.then` over `.success` and `.catch` over `.failure` - Angular 1.6 is removing those..

himdel commented:
> @himdel, moving the supportfeaturesmixin on the base class helps

Great news :)

So.. first of all, thanks for splitting the new changes into commits, much easier to review now! :bowing_man: 

But the bad news is... now I just don't see the Edit button anywhere, at all, for any floating ip, not in the list screen, nor in the detail screen. Did you add some check to hide it or something similar?

And another detail, the field marked "Floating IP address (optional)" seems to be the name field. So maybe the label is rather misleading, since it's not really for addresses ,but names :).

himdel commented on app/controllers/floating_ip_controller.rb:
```diff
@@ -14,4 +14,267 @@ def self.display_methods
+      task_id = @floating_ip.update_floating_ip_queue(session[:userid], options)
+
+      add_flash(_("Floating IP update failed: Task start failed: ID [%{id}]") %
+                {:id => task_id.to_s}, :error) unless task_id.kind_of?(Fixnum)
```
Fixnum should be Integer now (being fixed in https://github.com/ManageIQ/manageiq/pull/13191 for the rest)

gildub replied:
Yes it should, can't wait until we have all of those gone. I fixed it here.

himdel replied:
Where? ;)

himdel commented on app/controllers/floating_ip_controller.rb:
```diff
@@ -14,4 +14,267 @@ def self.display_methods
+      task_id = ems.create_floating_ip_queue(session[:userid], options)
+
+      add_flash(_("Floating IP creation failed: Task start failed: ID [%{id}]") %
+                {:id => task_id.to_s}, :error) unless task_id.kind_of?(Fixnum)
```
same, Fixnum should be Integer now (being fixed in https://github.com/ManageIQ/manageiq/pull/13191 for the rest)

himdel replied:
Uh, done where?

tzumainn commented:
@miq-bot add_label blocker
@miq-bot add_label bug

gildub commented on app/controllers/floating_ip_controller.rb:
```diff
@@ -14,4 +14,267 @@ def self.display_methods
+    when 'floating_ip_delete'
+      delete_floating_ips
+    when "floating_ip_edit"
+      checked_floating_ip_id = get_checked_floating_ip_id(params)
```
Yes, good DRY. Done

gildub commented on app/controllers/floating_ip_controller.rb:
```diff
@@ -14,4 +14,267 @@ def self.display_methods
+    available_networks = CloudNetwork.where(:ems_id => params[:id], :external_facing => true).find_each
+    available_networks.each do |network|
+      networks << { 'name' => network.name, 'id' => network.id }
+    end
```
Yes, absolutely, good poultry stuff, thanks!

himdel replied:
Oh, absolutely, I missed the array vs hash distinction, makes sense :)

gildub commented:
@himdel, 

The name of a floating ip is the floating address.
The edit (and delete) buttons were not showing up any more because the filter was wrong, I've fixed it.

himdel commented:
@gildub thanks!, going to re-review right now :)

> The name of a floating ip is the floating address.

Yeaah, possibly only in the openstack land.. 
![float](https://cloud.githubusercontent.com/assets/289743/21313352/5a43a840-c5e9-11e6-8704-5e8d19bfb884.png)

But I'm not blocking the PR for that, I guess that will change once you can edit other providers' floating ips too..

himdel commented:
So, @gildub I think this one is ready to get merged :+1:

But.. I'm seeing you marked a few Fixnum->Integers as fixed but haven't actually fixed them .. so I'm guessing you may have forgotten to push a commit here?

gildub commented:
@himdel, 

Sorry, yes the Fixnum -> Integer commit slipped out, I've restored it and rebased.

himdel commented:
> Sorry, yes the Fixnum -> Integer commit slipped out, I've restored it and rebased.

@gildub yeah, happens, thanks :) .. Just one more thing, can you please look at https://github.com/ManageIQ/manageiq/pull/12097/files#r93028604 ([line 146](https://github.com/ManageIQ/manageiq/pull/12097/files#diff-cb580b24d878da0531aed3049acfe8d4R146) looks like it should be used instead of [line 147](https://github.com/ManageIQ/manageiq/pull/12097/files#diff-cb580b24d878da0531aed3049acfe8d4R147)).

But if not for some reason, at least remove the commented out code please :).

gildub commented:
@himdel,

> But if not for some reason, at least remove the commented out code please :).

sorry I forgot about that one, it's fixed, using `find_by_id_filtered` and without any comments ;)

himdel commented:
@gildub Thanks, LGTM, the spec failures are probably all unrelated, I'll try to restart the tests once master ir OK, and get this is before the UI split..

simaishi commented:
Euwe backport details:

```
$ git log -1
commit fc719649568c5b0fadc90479c88f00625decb886
Author: Martin Hradil <himdel@seznam.cz>
Date:   Wed Dec 21 14:56:01 2016 +0000

    Merge pull request #12097 from gildub/network-provisioning-floating-ips-ui
    
    Floating IPs provisioning UI
    (cherry picked from commit 57b0a61fa1d9529582056123b68c47d7707a26ca)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1410588
```
